src/node/node_class.hpp

```diff
@@ -1195,123 +1187,87 @@ ObjectWrap<ClassType>::create_napi_property_descriptors(Napi::Env env, const Nap
-    if (!isRealmObjectClass) {
-        throw Napi::Error::New(env, "Creating instances by schema is supported for RealmObjectClass only");
-    }
+    std::string path = realm->config().path;
+    std::string schemaName = schema.name + ":" + std::to_string(realm->schema_version());
 
-    if (isRealmObjectClass && !internal) {
-        throw Napi::Error::New(
-            env, "RealmObjectClass requires an internal realm object when creating instances by schema");
-    }
+    auto& schemaObjects = s_schemaObjectTypes[path];
+    auto& schemaObject = schemaObjects[schemaName];
 
-    Napi::Object instance;
-    auto config = internal->realm()->config();
-    std::string path = config.path;
-    auto version = internal->realm()->schema_version();
-    std::string schemaName = schema.name + ":" + std::to_string(version);
-
-    std::unordered_map<std::string, SchemaObjectType*>* schemaObjects = nullptr;
-    if (!s_schemaObjectTypes.count(path)) {
-        // std::map<std::string, std::map<std::string, SchemaObjectType*>>
-        schemaObjects = new std::unordered_map<std::string, SchemaObjectType*>();
-        s_schemaObjectTypes.emplace(path, schemaObjects);
-    }
-    else {
-        schemaObjects = s_schemaObjectTypes.at(path);
-    }
 
+    auto ownKeys = Napi::Array::New(env);
     Napi::Function schemaObjectConstructor;
-    Napi::Symbol externalSymbol = ExternalSymbol;
+
     // if we are creating a RealmObject from schema with no user defined constructor
     if (constructor.IsEmpty()) {
-        // 1.Check by name if the constructor is already created for this RealmObject
-        if (!schemaObjects->count(schemaName)) {
-
-            // 2.Create the constructor
+        // 1.Check by name if the factory is already created for this RealmObject
+        //  XXX should we also check if `constructor` is empty?
+        if (!schemaObject.factory.IsEmpty())
+            return schemaObject.factory;
+
+
+        // 2.Create the constructor
+
+        // Create the RealmObject function.
+        // Weird syntax is the simplest way to make a function with the `name` property set to the value of a
+        // passed-in variable. This avoids the risk of users injecting code via the scema name.
+        schemaObjectConstructor = env.Global()
+                                      .Get("Function")
+                                      .As<Napi::Function>()
+                                      .New({
+                                          Napi::String::New(env, "name"), // argument name
+                                          Napi::String::New(env, "return {[name]: function () {}}[name];") // body
+                                      })
+                                      .As<Napi::Function>()
+                                      .Call({Napi::String::New(env, schema.name)})
+                                      .As<Napi::Function>();
```

Comment:
Couldn't we use something like https://github.com/nodejs/node-addon-api/blob/main/doc/function.md#new for that?

```suggestion
        schemaObjectConstructor = Napi::Function::New(env, [] {}, schema.name);
```